SwiftUI Picker - System photos picker

etoledom opened the pull request:
### Description

This PR implements the system photos picker for the Gravatar picker in SwiftUI.

**Notes:** 
- As a demo, I have implemented both add image buttons: in-grid and bottom CTA.
- We are only using the UIKit photos picker, and not the new SwiftUI PhotosPicker, since the later only supports Photo Library, doesn't support cropping, and the selection seems bugged (made for multi-select only, though it can return only one).
- The interface of the `ImagePickerView` is similar to the one of SwiftUI's PhotosPicker, to facilitate a replacement in the future.
- After the selected image is uploaded to the backend, we add the local image to the cache, to avoid an unnecessary download.

**Not included on this PR**:
- Selector between Camera and Library.
- Possibility to inject custom images picker / cropper. (do we want this in the first release?)

<img src="https://github.com/user-attachments/assets/1350c243-68c1-4b40-84cc-303077188ea0" width="40%">

### Testing Steps

- SwiftUI demo app.

etoledom committed:
Photo album picker

etoledom committed:
Merge branch 'trunk' into etoledom/swiftui-picker-pick-photo

etoledom committed:
Adding selected image to cache

etoledom committed:
Make UIImage extension for squared image

etoledom committed:
Fix unit test

etoledom committed:
Fix demo app build

pinarol commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift:
```diff
@@ -182,6 +183,69 @@ struct AvatarPickerView: View {
+}
+
+extension Color {
+    static let gravatarBlue = Color(uiColor: UIColor(red: 0.11, green: 0.31, blue: 0.77, alpha: 1.00))
```
This color seems like an important color. Let's either put this into the existing UIColor+DesignSystem extension or define a new one for swiftui like Color+DesignSystem put there.

etoledom replied:
Yup sounds good!

I can do this now or we do this when we implement dark mode (if it needs special handling with custom colors).

pinarol replied:
Let's do this now. Let's try to handle dark mode as we go. We can use this color for both light/dark modes I think. It didn't look weird when I tried.

pinarol commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerViewModel.swift:
```diff
@@ -93,6 +95,36 @@ class AvatarPickerViewModel: ObservableObject {
+            let newModel = AvatarImageModel(id: avatar.id, source: .remote(url: avatar.url))
+            add(newModel, replacing: localID)
+        } catch {
+            avatarsResult = .failure(error)
```
Let's not use avatarsResult for upload result. We would want to be able to distinguish the failure of avatar fetching from failure of avatar upload.

pinarol replied:
And I realized we don't have any error state in the designs for when the avatar upload fails. I think it is missing atm. But still let's have a different result object for upload.

etoledom replied:
True. Since we don't have yet a proper error handling (at the time I was writing this PR), I didn't put much time on this. 

I'd work on this proper error handling on a separate PR once we have the design.

etoledom replied:
We will probably have a per-avatar error state to show in the grid, but let's wait for the proper design before implementing something like this.

pinarol replied:
okay then let's at least remove that line add a TODO since it is almost certain that we'll have a quite different error handling for avatar upload.

pinarol commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift:
```diff
@@ -3,45 +3,29 @@ import SwiftUI
-            errorMessages()
-
-            profileView()
+        VStack {
```
Just curious. Why was the extra layer of VStack necessary?

etoledom replied:
To keep the CTA button pinned at the bottom, outside the scroll view.

pinarol commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerViewModel.swift:
```diff
@@ -93,6 +95,36 @@ class AvatarPickerViewModel: ObservableObject {
+        let service = AvatarService()
+        do {
+            let avatar = try await service.upload(image, accessToken: authToken)
+            await ImageCache.shared.setEntry(.ready(squareImage), for: avatar.url)
```
Do we ever fetch this avatar with the same `avatar.url` again? I feel like next time we try to fetch the avatar, the url will be different because of other query params like size and stuff. So this entry won't be used, no?

etoledom replied:
Since we use the same `Avatar` model, the `avatar.url` is the same than the remote avatars used in the grid. It already includes the base url and size param. This is why I decided to use the same model (plus the image structure returned by the upload endpoint is the same than the one returned by the avatar list endpoint)

We will see how this will be handled in the open-api spec, but I'd expect it to be the same structure reused.

pinarol reviewed (commented):
Looks nice overall but I failed to make a successful upload, getting an error from server. Trying to figure out why.

etoledom committed:
Avoid mistake of uploading a non-squared image

etoledom commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift:
```diff
@@ -182,6 +183,69 @@ struct AvatarPickerView: View {
     }
 }
 
+struct UploadImageCTAButtonView: View {
+    public var body: some View {
+        Text("Upload image")
+            .font(.callout).fontWeight(.bold)
+            .frame(maxWidth: .infinity)
+            .foregroundColor(.white)
+            .padding()
+            .background(RoundedRectangle(cornerRadius: 4).fill(Color.gravatarBlue))
+    }
+}
```
I'm not sure if there's a benefit from having these views as separated structs, or having them as view builder functions.

I had these `UploadImageCTAButtonView`, `ActivityIndicatorView`, `PlusButtonView`, as functions before, but then I decided to make separated structs to make the main `AvatarPickerView` smaller.

Maybe one benefit could be that we could have these structs in a separate file with their own previews, and/or snapshot tests.

pinarol replied:
The only benefit would be to easily reuse them from other views, currently we don't have such other views so the benefit isn't here yet. But I don't mind having them as separate views. Also, why don't we put them in separate files also?

etoledom commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift:
```diff
@@ -182,6 +183,69 @@ struct AvatarPickerView: View {
+    static let gravatarBlue = Color(uiColor: UIColor(red: 0.11, green: 0.31, blue: 0.77, alpha: 1.00))
+}
+
+private enum Constants {
+    static let horizontalPadding: CGFloat = .DS.Padding.double
```
I don't particularly like having these constants declared globally, even though is private (I prefer them to be constrained inside the struct) but it was necessary to share with the other new view structs.

Any special thought?

pinarol replied:
Ideally, if we are going to have separate reusable views, they shouldn't use the constants that belong to AvatarPickerView. We should pass these numbers as input parameters to them and put Constants back into AvatarPickerView imho because those Constants only mean something in that specific context. WDYT?

etoledom committed:
Reusable `CTAButtonView`

etoledom committed:
Moving gravatarBlue to DesignSystem

etoledom committed:
Move views to their files

etoledom committed:
Rename OverlayActivityIndicatorView

etoledom committed:
Lint

etoledom committed:
Remove incorrect error handling of avatar upload

etoledom commented:
@pinarol - Ready for another look 👍 

- Moved the color to the design system declaration
  - The designs has the same blue for both light and dark modes.
- Moved the new struct views to their own files with previews.
- The `CTAButtonView` has variable title.
- Constants are back to be bounded to AvatarPickerView.
- Removed the wrong error handling from avatar upload.
- Few other small things.

pinarol commented on Sources/GravatarUI/SwiftUI/AvatarPicker/AvatarPickerView.swift:
```diff
@@ -112,13 +126,17 @@ struct AvatarPickerView: View {
                     loadingView: {
-                        ProgressView()
-                            .progressViewStyle(CircularProgressViewStyle())
+                        OverlayActivityIndicatorView()
```
We don't need the overlay part for the downloading state we just need the progress view part. The overlay is messing up with the background color. Can we keep the plain old ProgressView here?

pinarol reviewed (approved):
Looking good 🎉 Just one more tiny comment but I don't need to check again I think.

etoledom committed:
Roll back activity indicator on Avatar Image
